nomt/src/io/page_pool.rs

```diff
@@ -158,44 +189,63 @@ impl PagePool {
         self.inner.freelist.write().push(page);
     }
 
-    fn grow(&self, freelist: &mut Vec<Page>) -> Page {
-        assert!(freelist.is_empty());
+    fn grow(&self, freelist_guard: &mut RwLockWriteGuard<Vec<Page>>) -> Page {
+        assert!(freelist_guard.is_empty());
+
+        // First step is to allocate a new region.
         let region_ptr = unsafe {
             libc::mmap(
                 std::ptr::null_mut(),
-                REGION_SIZE,
+                REGION_BYTE_SIZE,
                 libc::PROT_READ | libc::PROT_WRITE,
                 libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
                 /* fd */ -1,
                 /* offset */ 0,
             )
         };
+        assert!(!region_ptr.is_null());
         if region_ptr == libc::MAP_FAILED {
             panic!("Failed to allocate memory");
         }
-        let mut regions = self.inner.regions.write();
-        regions.push(region_ptr);
 
-        let region: u32 = ((regions.len() - 1) as u32) << 16;
-        for slot in 0..=REGION_SLOT_MASK {
-            freelist.push(Page(PageIndex(slot | region)));
+        // Next, we need to store the region pointer in the regions array.
+        //
+        // We store the pointer in the regions array before incrementing n_regions. This is not
+        // strictly necessary, because the freelist is empty and no page can refer to the new region
+        // yet. Likewise, drop cannot happen during this operation. We still do it in this order
+        // to just err on the safe side and avoid any potential issues.
+        let region_ix = self.inner.n_regions.load(Ordering::Relaxed);
+        self.inner.regions[region_ix as usize].store(region_ptr as *mut u8, Ordering::Release);
+        self.inner.n_regions.fetch_add(1, Ordering::Relaxed);
```

Comment:
note: this is protected by the write guard on the free list but if it were lock-free we'd want to claim the region index with a CAS rather than read followed by fetch-add.

Reply:
yep, I don't think that matters much as we disccused and I also updated the comment to reflect that.